--- /var/ida/nextcloud-new/core/templates/layout.guest.php	2024-11-05 08:30:08.259702233 +0000
+++ /var/ida/nextcloud-old/core/templates/layout.guest.php.original	2024-02-20 10:11:39.209443037 +0000
@@ -1,12 +1,5 @@
-<?php
-/**
- * SPDX-FileCopyrightText: 2016-2024 Nextcloud GmbH and Nextcloud contributors
- * SPDX-FileCopyrightText: 2011-2016 ownCloud, Inc.
- * SPDX-License-Identifier: AGPL-3.0-only
- */
-?>
 <!DOCTYPE html>
-<html class="ng-csp" data-placeholder-focus="false" lang="<?php p($_['language']); ?>" data-locale="<?php p($_['locale']); ?>" translate="no" >
+<html class="ng-csp" data-placeholder-focus="false" lang="<?php p($_['language']); ?>" data-locale="<?php p($_['locale']); ?>" >
 	<head
 <?php if ($_['user_uid']) { ?>
 	data-user="<?php p($_['user_uid']); ?>" data-user-displayname="<?php p($_['user_displayname']); ?>"
@@ -14,55 +7,52 @@
  data-requesttoken="<?php p($_['requesttoken']); ?>">
 		<meta charset="utf-8">
 		<title>
-			<?php
-				p(!empty($_['pageTitle']) ? $_['pageTitle'] . ' – ' : '');
-p($theme->getTitle());
-?>
+		<?php p($theme->getTitle()); ?>
 		</title>
-		<meta name="csp-nonce" nonce="<?php p($_['cspNonce']); /* Do not pass into "content" to prevent exfiltration */ ?>">
-		<meta name="viewport" content="width=device-width, initial-scale=1.0, minimum-scale=1.0<?php if (isset($_['viewport_maximum_scale'])) {
-			p(', maximum-scale=' . $_['viewport_maximum_scale']);
-		} ?>">
+		<meta http-equiv="X-UA-Compatible" content="IE=edge">
+		<meta name="viewport" content="width=device-width, initial-scale=1.0, minimum-scale=1.0">
 		<?php if ($theme->getiTunesAppId() !== '') { ?>
 		<meta name="apple-itunes-app" content="app-id=<?php p($theme->getiTunesAppId()); ?>">
 		<?php } ?>
 		<meta name="theme-color" content="<?php p($theme->getColorPrimary()); ?>">
-		<link rel="icon" href="<?php print_unescaped(image_path('core', 'favicon.ico')); /* IE11+ supports png */ ?>">
-		<link rel="apple-touch-icon" href="<?php print_unescaped(image_path('core', 'favicon-touch.png')); ?>">
-		<link rel="mask-icon" sizes="any" href="<?php print_unescaped(image_path('core', 'favicon-mask.svg')); ?>" color="<?php p($theme->getColorPrimary()); ?>">
-		<link rel="manifest" href="<?php print_unescaped(image_path('core', 'manifest.json')); ?>" crossorigin="use-credentials">
+		<link rel="icon" href="<?php print_unescaped(image_path('', 'favicon.ico')); /* IE11+ supports png */ ?>">
+		<link rel="apple-touch-icon" href="<?php print_unescaped(image_path('', 'favicon-touch.png')); ?>">
+		<link rel="mask-icon" sizes="any" href="<?php print_unescaped(image_path('', 'favicon-mask.svg')); ?>" color="<?php p($theme->getColorPrimary()); ?>">
+		<link rel="manifest" href="<?php print_unescaped(image_path('', 'manifest.json')); ?>">
 		<?php emit_css_loading_tags($_); ?>
 		<?php emit_script_loading_tags($_); ?>
 		<?php print_unescaped($_['headers']); ?>
 	</head>
 	<body id="<?php p($_['bodyid']);?>">
 		<?php include 'layout.noscript.warning.php'; ?>
-		<?php include 'layout.initial-state.php'; ?>
+		<?php foreach ($_['initialStates'] as $app => $initialState) { ?>
+			<input type="hidden" id="initial-state-<?php p($app); ?>" value="<?php p(base64_encode($initialState)); ?>">
+		<?php }?>
 		<div class="wrapper">
 			<div class="v-align">
 				<?php if ($_['bodyid'] === 'body-login'): ?>
-					<header>
+					<header role="banner">
 						<div id="header">
-							<div class="logo"></div>
+							<div class="logo">
+								<h1 class="hidden-visually">
+									<?php p($theme->getName()); ?>
+								</h1>
+								<?php if (\OC::$server->getConfig()->getSystemValue('installed', false)
+									&& \OC::$server->getConfig()->getAppValue('theming', 'logoMime', false)): ?>
+									<img src="<?php p($theme->getLogo()); ?>"/>
+								<?php endif; ?>
+							</div>
 						</div>
 					</header>
 				<?php endif; ?>
 				<main>
-					<h1 class="hidden-visually">
-						<?php p($theme->getName()); ?>
-					</h1>
 					<?php print_unescaped($_['content']); ?>
 				</main>
 			</div>
 		</div>
-		<?php
-		$longFooter = $theme->getLongFooter();
-?>
-		<footer class="guest-box <?php if ($longFooter === '') {
-			p('hidden');
-		} ?>">
+		<footer role="contentinfo">
 			<p class="info">
-				<?php print_unescaped($longFooter); ?>
+				<?php print_unescaped($theme->getLongFooter()); ?>
 			</p>
 		</footer>
 	</body>
