--- /var/ida/nextcloud-new/apps/settings/templates/settings/personal/security/password.php	2024-11-05 08:30:08.897704441 +0000
+++ /var/ida/nextcloud-old/apps/settings/templates/settings/personal/security/password.php	2024-02-20 10:11:37.866438484 +0000
@@ -1,11 +1,24 @@
 <?php
 /**
- * SPDX-FileCopyrightText: 2017 Nextcloud GmbH and Nextcloud contributors
- * SPDX-License-Identifier: AGPL-3.0-or-later
+ * @copyright Copyright (c) 2017 Arthur Schiwon <blizzz@arthur-schiwon.de>
+ *
+ * @author Arthur Schiwon <blizzz@arthur-schiwon.de>
+ *
+ * @license GNU AGPL version 3 or any later version
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
  */
 
-if ($_['passwordChangeSupported']) {
-	\OCP\Util::addScript('settings', 'vue-settings-personal-password');
-}
 ?>
-<div id="security-password"></div>
