--- /var/ida/nextcloud-old/lib/private/Config.php	2024-02-20 10:11:39.275443260 +0000
+++ /var/ida/nextcloud-old/lib/private/Config.php.original	2024-02-20 10:11:39.275443260 +0000
@@ -66,9 +66,6 @@
 		$this->configFilePath = $this->configDir.$fileName;
 		$this->configFileName = $fileName;
 		$this->readData();
-		// Force ISO 8610 UTC log timestamps
-		$this->set('logtimezone','UTC');
-		$this->set('logdateformat', 'Y-m-d\\TH:i:s\\Z');
 	}
 
 	/**
