--- /var/ida/nextcloud-new/apps/settings/templates/settings/personal/security/twofactor.php	2024-11-05 08:30:08.897704441 +0000
+++ /var/ida/nextcloud-old/apps/settings/templates/settings/personal/security/twofactor.php.original	2024-02-20 10:11:37.867438488 +0000
@@ -1,8 +1,25 @@
 <?php
 declare(strict_types=1);
 /**
- * SPDX-FileCopyrightText: 2020 Nextcloud GmbH and Nextcloud contributors
- * SPDX-License-Identifier: AGPL-3.0-or-later
+ * @copyright Copyright (c) 2020, Roeland Jago Douma <roeland@famdouma.nl>
+ *
+ * @author Roeland Jago Douma <roeland@famdouma.nl>
+ *
+ * @license GNU AGPL version 3 or any later version
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
  */
 
 ?>
@@ -23,11 +40,19 @@
 			$provider = $data['provider'];
 		//Handle 2FA provider icons and theme
 		if ($provider instanceof \OCP\Authentication\TwoFactorAuth\IProvidesIcons) {
+				if ($_['themedark']) {
+					$icon = $provider->getLightIcon();
+				} else {
 			$icon = $provider->getDarkIcon();
+				}
 			//fallback icon if the 2factor provider doesn't provide an icon.
 		} else {
+				if ($_['themedark']) {
+					$icon = image_path('core', 'actions/password-white.svg');
+				} else {
 			$icon = image_path('core', 'actions/password.svg');
 		}
+			}
 		/** @var \OCP\Authentication\TwoFactorAuth\IPersonalProviderSettings $settings */
 		$settings = $data['settings'];
 		?>
