--- /var/ida/nextcloud-new/apps/files_sharing/appinfo/routes.php	2024-11-05 08:30:08.957704649 +0000
+++ /var/ida/nextcloud-old/apps/files_sharing/appinfo/routes.php	2024-02-20 10:11:36.944435359 +0000
@@ -1,9 +1,31 @@
 <?php
 /**
- * SPDX-FileCopyrightText: 2016-2024 Nextcloud GmbH and Nextcloud contributors
- * SPDX-FileCopyrightText: 2016 ownCloud, Inc.
- * SPDX-License-Identifier: AGPL-3.0-only
+ * @copyright Copyright (c) 2016, ownCloud, Inc.
+ *
+ * @author Björn Schießle <bjoern@schiessle.org>
+ * @author Christoph Wurst <christoph@winzerhof-wurst.at>
+ * @author Joas Schilling <coding@schilljs.com>
+ * @author John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>
+ * @author Lukas Reschke <lukas@statuscode.ch>
+ * @author Maxence Lange <maxence@artificial-owl.com>
+ * @author Roeland Jago Douma <roeland@famdouma.nl>
+ *
+ * @license AGPL-3.0
+ *
+ * This code is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License, version 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License, version 3,
+ * along with this program. If not, see <http://www.gnu.org/licenses/>
+ *
  */
+
 return [
 	'resources' => [
 		'ExternalShares' => ['url' => '/api/externalShares'],
@@ -62,16 +84,6 @@
 			'verb' => 'PUT',
 		],
 		[
-			'name' => 'Settings#setUserShareFolder',
-			'url' => '/settings/shareFolder',
-			'verb' => 'PUT',
-		],
-		[
-			'name' => 'Settings#resetUserShareFolder',
-			'url' => '/settings/shareFolder',
-			'verb' => 'DELETE',
-		],
-		[
 			'name' => 'Accept#accept',
 			'url' => '/accept/{shareId}',
 			'verb' => 'GET',
@@ -117,11 +129,6 @@
 			'verb' => 'DELETE',
 		],
 		[
-			'name' => 'ShareAPI#sendShareEmail',
-			'url' => '/api/v1/shares/{id}/send-email',
-			'verb' => 'POST',
-		],
-		[
 			'name' => 'ShareAPI#acceptShare',
 			'url' => '/api/v1/shares/pending/{id}',
 			'verb' => 'POST',
@@ -141,7 +148,6 @@
 		],
 		/*
 		 * OCS Sharee API
-		 */
 		[
 			'name' => 'ShareesAPI#search',
 			'url' => '/api/v1/sharees',
@@ -152,6 +158,7 @@
 			'url' => '/api/v1/sharees_recommended',
 			'verb' => 'GET',
 		],
+		 */
 		/*
 		 * Remote Shares
 		 */
