--- /var/ida/nextcloud-old/lib/private/Authentication/Token/PublicKeyTokenProvider.php	2024-02-20 10:11:39.269443240 +0000
+++ /var/ida/nextcloud-old/lib/private/Authentication/Token/PublicKeyTokenProvider.php.original	2024-02-20 10:11:39.269443240 +0000
@@ -417,10 +417,6 @@
 		// Update the password for all tokens
 		$tokens = $this->mapper->getTokenByUser($uid);
 		foreach ($tokens as $t) {
-			// But, do not add a password to passwordless tokens.
-			if (is_null($t->getPassword())) {
-				continue;
-			}
 			$publicKey = $t->getPublicKey();
 			$t->setPassword($this->encryptPassword($password, $publicKey));
 			$t->setPasswordInvalid(false);
