--- /var/ida/nextcloud-old/apps/files/lib/Controller/ViewController.php	2024-02-20 10:11:36.748434694 +0000
+++ /var/ida/nextcloud-old/apps/files/lib/Controller/ViewController.php.original	2024-02-20 10:11:36.748434694 +0000
@@ -307,7 +307,7 @@
 		$params['defaultFileSortingDirection'] = $this->config->getUserValue($user, 'files', 'file_sorting_direction', 'asc');
 		$params['showgridview'] = $this->config->getUserValue($user, 'files', 'show_grid', false);
 		$params['isIE'] = \OCP\Util::isIE();
-		$showHidden = true;
+		$showHidden = (bool) $this->config->getUserValue($this->userSession->getUser()->getUID(), 'files', 'show_hidden', false);
 		$params['showHiddenFiles'] = $showHidden ? 1 : 0;
 		$cropImagePreviews = (bool) $this->config->getUserValue($this->userSession->getUser()->getUID(), 'files', 'crop_image_previews', true);
 		$params['cropImagePreviews'] = $cropImagePreviews ? 1 : 0;
