--- /var/ida/nextcloud-new/lib/public/AppFramework/Http/ContentSecurityPolicy.php	2024-11-05 08:30:09.572706778 +0000
+++ /var/ida/nextcloud-old/lib/public/AppFramework/Http/ContentSecurityPolicy.php	2024-04-04 12:24:01.225437639 +0000
@@ -1,10 +1,29 @@
 <?php
-
 /**
- * SPDX-FileCopyrightText: 2016-2024 Nextcloud GmbH and Nextcloud contributors
- * SPDX-FileCopyrightText: 2016 ownCloud, Inc.
- * SPDX-License-Identifier: AGPL-3.0-only
+ * @copyright Copyright (c) 2016, ownCloud, Inc.
+ *
+ * @author Lukas Reschke <lukas@statuscode.ch>
+ * @author Morris Jobke <hey@morrisjobke.de>
+ * @author Roeland Jago Douma <roeland@famdouma.nl>
+ * @author sualko <klaus@jsxc.org>
+ * @author Thomas Citharel <nextcloud@tcit.fr>
+ *
+ * @license AGPL-3.0
+ *
+ * This code is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License, version 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License, version 3,
+ * along with this program. If not, see <http://www.gnu.org/licenses/>
+ *
  */
+
 namespace OCP\AppFramework\Http;
 
 /**
@@ -26,15 +45,21 @@
 	protected $inlineScriptAllowed = false;
 	/** @var bool Whether eval in JS scripts is allowed */
 	protected $evalScriptAllowed = false;
-	/** @var bool Whether WebAssembly compilation is allowed */
-	protected ?bool $evalWasmAllowed = false;
-	/** @var bool Whether strict-dynamic should be set */
-	protected $strictDynamicAllowed = false;
-	/** @var bool Whether strict-dynamic should be set for 'script-src-elem' */
-	protected $strictDynamicAllowedOnScripts = true;
 	/** @var array Domains from which scripts can get loaded */
 	protected $allowedScriptDomains = [
 		'\'self\'',
+		'metrics.fairdata.fi',
+		'metrics.fd-test.csc.fi',
+		'sso.fairdata.fi',
+		'sso.demo.fairdata.fi',
+		'sso.fd-test.csc.fi',
+		'sso.fd-stable.csc.fi',
+		'sso.fd-staging.csc.fi',
+		'sso.fd-rework.csc.fi',
+		'sso.fd-dev.csc.fi',
+		'ps.fd-dev.csc.fi:4443',
+		'psdev.fd-dev.csc.fi:4443',
+		'psdev-nc21.fd-dev.csc.fi:4443',
 	];
 	/**
 	 * @var bool Whether inline CSS is allowed
@@ -45,6 +70,18 @@
 	/** @var array Domains from which CSS can get loaded */
 	protected $allowedStyleDomains = [
 		'\'self\'',
+		'metrics.fairdata.fi',
+		'metrics.fd-test.csc.fi',
+		'sso.fairdata.fi',
+		'sso.demo.fairdata.fi',
+		'sso.fd-test.csc.fi',
+		'sso.fd-stable.csc.fi',
+		'sso.fd-staging.csc.fi',
+		'sso.fd-rework.csc.fi',
+		'sso.fd-dev.csc.fi',
+		'ps.fd-dev.csc.fi:4443',
+		'psdev.fd-dev.csc.fi:4443',
+		'psdev-nc21.fd-dev.csc.fi:4443',
 	];
 	/** @var array Domains from which images can get loaded */
 	protected $allowedImageDomains = [
@@ -55,6 +92,8 @@
 	/** @var array Domains to which connections can be done */
 	protected $allowedConnectDomains = [
 		'\'self\'',
+		'metrics.fairdata.fi',
+		'metrics.fd-test.csc.fi',
 	];
 	/** @var array Domains from which media elements can be loaded */
 	protected $allowedMediaDomains = [
