--- /var/ida/nextcloud-new/apps/files/js/mainfileinfodetailview.js	2024-11-05 08:30:09.001704801 +0000
+++ /var/ida/nextcloud-old/apps/files/js/mainfileinfodetailview.js	2024-02-20 10:11:36.711434569 +0000
@@ -1,7 +1,11 @@
-/**
- * SPDX-FileCopyrightText: 2016-2024 Nextcloud GmbH and Nextcloud contributors
- * SPDX-FileCopyrightText: 2015-2016 ownCloud, Inc.
- * SPDX-License-Identifier: AGPL-3.0-or-later
+/*
+ * Copyright (c) 2015
+ *
+ * This file is licensed under the Affero General Public License version 3
+ * or later.
+ *
+ * See the COPYING-README file.
+ *
  */
 
 (function() {
@@ -39,8 +43,8 @@
 		events: {
 			'click a.action-favorite': '_onClickFavorite',
 			'click a.action-default': '_onClickDefaultAction',
-			'click a.permalink': '_onClickPermalink',
-			'focus .permalink-field>input': '_onFocusPermalink'
+			//'click a.permalink': '_onClickPermalink',
+			//'focus .permalink-field>input': '_onFocusPermalink'
 		},
 
 		template: function(data) {
@@ -63,9 +67,20 @@
 		},
 
 		_setupClipboard: function() {
+			return;
 			var clipboard = new Clipboard('.permalink');
 			clipboard.on('success', function(e) {
-				OC.Notification.show(t('files', 'Direct link was copied (only works for people who have access to this file/folder)'), {type: 'success'});
+				var $el = $(e.trigger);
+				$el.tooltip('hide')
+					.attr('data-original-title', t('core', 'Copied!'))
+					.tooltip('fixTitle')
+					.tooltip({placement: 'bottom', trigger: 'manual'})
+					.tooltip('show');
+				_.delay(function() {
+					$el.tooltip('hide');
+					$el.attr('data-original-title', t('files', 'Copy direct link (only works for users who have access to this file/folder)'))
+						.tooltip('fixTitle');
+				}, 3000);
 			});
 			clipboard.on('error', function(e) {
 				var $row = this.$('.permalink-field');
@@ -77,11 +92,12 @@
 		},
 
 		_onClickPermalink: function(e) {
-			e.preventDefault();
+			//e.preventDefault();
 			return;
 		},
 
 		_onFocusPermalink: function() {
+			return;
 			this.$('.permalink-field>input').select();
 		},
 
@@ -140,8 +156,7 @@
 				var availableActions = this._fileActions.get(
 					this.model.get('mimetype'),
 					this.model.get('type'),
-					this.model.get('permissions'),
-					this.model.get('name')
+					this.model.get('permissions')
 				);
 				var hasFavoriteAction = 'Favorite' in availableActions;
 				this.$el.html(this.template({
@@ -152,18 +167,17 @@
 					path: this.model.get('path'),
 					hasSize: this.model.has('size'),
 					sizeLabel: t('files', 'Size'),
-					size: OC.Util.humanFileSize(this.model.get('size'), true, false),
+					size: OC.Util.humanFileSize(this.model.get('size'), true),
 					altSize: n('files', '%n byte', '%n bytes', this.model.get('size')),
 					dateLabel: t('files', 'Modified'),
 					altDate: OC.Util.formatDate(this.model.get('mtime')),
 					timestamp: this.model.get('mtime'),
 					date: OC.Util.relativeModifiedDate(this.model.get('mtime')),
 					hasFavoriteAction: hasFavoriteAction,
-					// TRANSLATORS "Favour" is a verb
-					starAltText: isFavorite ? t('files', 'Favored') : t('files', 'Favor'),
+					starAltText: isFavorite ? t('files', 'Favorited') : t('files', 'Favorite'),
 					starClass: isFavorite ? 'icon-starred' : 'icon-star',
-					permalink: this._makePermalink(this.model.get('id')),
-					permalinkTitle: t('files', 'Copy direct link (only works for people who have access to this file/folder)')
+					//permalink: this._makePermalink(this.model.get('id')),
+					//permalinkTitle: t('files', 'Copy direct link (only works for users who have access to this file/folder)')
 				}));
 
 				// TODO: we really need OC.Previews
@@ -179,6 +193,7 @@
 					}
 					$iconDiv.css('background-image', 'url("' + iconUrl + '")');
 				}
+				this.$el.find('[title]').tooltip({placement: 'bottom'});
 			} else {
 				this.$el.empty();
 			}
