--- /var/ida/nextcloud-old/lib/private/AppFramework/Middleware/Security/PasswordConfirmationMiddleware.php	2024-02-20 10:11:39.262443216 +0000
+++ /var/ida/nextcloud-old/lib/private/AppFramework/Middleware/Security/PasswordConfirmationMiddleware.php.original	2024-02-20 10:11:39.262443216 +0000
@@ -70,9 +70,6 @@
 	 */
 	public function beforeController($controller, $methodName) {
 		if ($this->reflector->hasAnnotation('PasswordConfirmationRequired')) {
-			// IDA MOD: Because we use the SSO, we never require password confirmation for actions 
-			// becausee Nextcloud doesn't have user passwords for all authentication methods.
-			return;
 			$user = $this->userSession->getUser();
 			$backendClassName = '';
 			if ($user !== null) {
