--- /var/ida/nextcloud-old/lib/private/EventDispatcher/GenericEventWrapper.php	2024-02-20 10:11:39.283443287 +0000
+++ /var/ida/nextcloud-old/lib/private/EventDispatcher/GenericEventWrapper.php.original	2024-02-20 10:11:39.283443287 +0000
@@ -52,9 +52,6 @@
 	}
 
 	private function log() {
-		// Do not log any deprecation notices at all.
-		return;
-
 		if ($this->deprecationNoticeLogged) {
 			return;
 		}
