--- /var/ida/nextcloud-old/lib/private/App/AppManager.php	2024-02-20 10:11:39.256443196 +0000
+++ /var/ida/nextcloud-old/lib/private/App/AppManager.php.original	2024-02-20 10:11:39.256443196 +0000
@@ -578,28 +578,7 @@
 			}
 			$content = json_decode(file_get_contents($shippedJson), true);
 			$this->shippedApps = $content['shippedApps'];
-			$alwaysEnabled = $content['alwaysEnabled'];
-			# Absolutely ensure that the IDA apps are included in the set of apps that are always enabled
-			if (!in_array("ida", $alwaysEnabled)) {
-			    array_push ($alwaysEnabled, "ida");
-			}
-			if (!in_array("idafirstrunwizard", $alwaysEnabled)) {
-			    array_push ($alwaysEnabled, "idafirstrunwizard");
-			}
-			$this->alwaysEnabled = $alwaysEnabled;
-		}
-		else {
-			# Absolutely ensure that the IDA apps are included in the set of apps that are always enabled
-			if ( (!in_array("ida", $this->alwaysEnabled)) || (! in_array("idafirstrunwizard", $this->alwaysEnabled)) ) {
-			    $alwaysEnabled = $this->alwaysEnabled;
-			    if (!in_array("ida", $alwaysEnabled)) {
-			        array_push ($alwaysEnabled, "ida");
-			    }
-			    if (!in_array("idafirstrunwizard", $alwaysEnabled)) {
-			        array_push ($alwaysEnabled, "idafirstrunwizard");
-			    }
-			    $this->alwaysEnabled = $alwaysEnabled;
-			}
+			$this->alwaysEnabled = $content['alwaysEnabled'];
 		}
 	}
 
