--- /var/ida/nextcloud-old/core/l10n/fi.js	2024-02-20 10:11:39.005442345 +0000
+++ /var/ida/nextcloud-old/core/l10n/fi.js.original	2024-02-20 10:11:39.005442345 +0000
@@ -1,9 +1,6 @@
 OC.L10N.register(
     "core",
     {
-    "Enter search terms here..." : "Kirjoita hakutermit tähän...",
-    "Note to recipient" : "Viesti vastaanottajalle",
-    "Security" : "Turvallisuus",
     "Please select a file." : "Valitse tiedosto.",
     "File is too big" : "Tiedosto on liian suuri",
     "The selected file is not an image." : "Valittu tiedosto ei ole kuva.",
